File: frontend/src/api/session.ts
import { fetchApi } from './client';
import { AgentKey, ConversationMeta } from '../types/store';

interface BookRef {
    kind: 'book_name' | 'book_id';
    value: string;
}

export interface ConversationRecordPayload {
    id: string;
    ts: string;
    role: 'user' | 'assistant' | 'system';
    text: string;
    conversation_id: string;
    upstream_conversation_id?: string | null;
    active_file: string;
    status: 'streaming' | 'done' | 'error' | 'interrupted';
}

export interface ConversationContextResponse {
    status: 'success';
    book_id: string;
    agent_key: AgentKey;
    active_conversation_id: string | null;
    conversation_id: string | null;
    upstream_conversation_id: string | null;
    conversations: ConversationMeta[];
    messages: ConversationRecordPayload[];
}

function getBookQuery(bookRef: BookRef): string {
    return bookRef.kind === 'book_name'
        ? `book_name=${encodeURIComponent(bookRef.value)}`
        : `book_id=${encodeURIComponent(bookRef.value)}`;
}

export async function fetchConversationContext(
    bookRef: BookRef,
    agentKey: AgentKey,
    conversationId?: string | null,
): Promise<ConversationContextResponse> {
    const conversationQuery = conversationId ? `&conversation_id=${encodeURIComponent(conversationId)}` : '';
    return fetchApi<ConversationContextResponse>(
        `/api/conversations/context?${getBookQuery(bookRef)}&agent=${encodeURIComponent(agentKey)}${conversationQuery}`
    );
}

export async function createConversation(
    bookRef: BookRef,
    agentKey: AgentKey,
    activeFile: string,
    title?: string,
): Promise<ConversationContextResponse> {
    return fetchApi<ConversationContextResponse>('/api/conversations/create', {
        method: 'POST',
        body: JSON.stringify({
            [bookRef.kind]: bookRef.value,
            agent: agentKey,
            active_file: activeFile,
            title,
        }),
    });
}

export async function activateConversation(
    bookRef: BookRef,
    agentKey: AgentKey,
    conversationId: string,
): Promise<ConversationContextResponse> {
    return fetchApi<ConversationContextResponse>('/api/conversations/activate', {
        method: 'POST',
        body: JSON.stringify({
            [bookRef.kind]: bookRef.value,
            agent: agentKey,
            conversation_id: conversationId,
        }),
    });
}

export async function renameConversation(
    bookRef: BookRef,
    agentKey: AgentKey,
    conversationId: string,
    title: string,
): Promise<ConversationContextResponse> {
    return fetchApi<ConversationContextResponse>('/api/conversations/rename', {
        method: 'POST',
        body: JSON.stringify({
            [bookRef.kind]: bookRef.value,
            agent: agentKey,
            conversation_id: conversationId,
            title,
        }),
    });
}

export async function archiveConversation(
    bookRef: BookRef,
    agentKey: AgentKey,
    conversationId: string,
): Promise<ConversationContextResponse> {
    return fetchApi<ConversationContextResponse>('/api/conversations/archive', {
        method: 'POST',
        body: JSON.stringify({
            [bookRef.kind]: bookRef.value,
            agent: agentKey,
            conversation_id: conversationId,
        }),
    });
}

export async function deleteConversation(
    bookRef: BookRef,
    agentKey: AgentKey,
    conversationId: string,
): Promise<ConversationContextResponse> {
    return fetchApi<ConversationContextResponse>('/api/conversations/delete', {
        method: 'POST',
        body: JSON.stringify({
            [bookRef.kind]: bookRef.value,
            agent: agentKey,
            conversation_id: conversationId,
        }),
    });
}

File: novel_git_server/agents/session.py
from __future__ import annotations

from typing import Callable

from flask import Blueprint, jsonify, request

from utils.session_runtime import (
    activate_conversation,
    archive_conversation,
    create_conversation,
    delete_conversation,
    get_agent_context,
    rename_conversation,
    route_agent_to_session_agent,
)


def create_blueprint(
    *,
    dev_repo_root: str,
    require_book_id: Callable[[dict | None], tuple[str | None, tuple | None]],
    json_error: Callable[[str, str, int], tuple],
) -> Blueprint:
    bp = Blueprint("session", __name__)

    @bp.get("/api/conversations/context")
    def get_conversation_context():
        book_id, book_err = require_book_id(None)
        if book_err:
            return book_err

        raw_agent = request.args.get("agent", "")
        try:
            agent_key = route_agent_to_session_agent(raw_agent)
        except ValueError:
            return json_error("INVALID_AGENT", f"unsupported agent: {raw_agent}", 400)

        raw_conversation_id = request.args.get("conversation_id")
        conversation_id = raw_conversation_id.strip() if isinstance(raw_conversation_id, str) and raw_conversation_id.strip() else None

        payload = get_agent_context(
            dev_repo_root,
            book_id=book_id,
            agent_key=agent_key,
            conversation_id=conversation_id,
        )
        return jsonify({"status": "success", **payload}), 200

    @bp.post("/api/conversations/create")
    def create_conversation_route():
        payload = request.get_json(silent=True) or {}
        book_id, book_err = require_book_id(payload)
        if book_err:
            return book_err
        raw_agent = payload.get("agent", "")
        raw_active_file = payload.get("active_file", "")
        raw_title = payload.get("title")
        if not isinstance(raw_active_file, str) or not raw_active_file.strip():
            return json_error("INVALID_PAYLOAD", "active_file is required", 400)
        try:
            agent_key = route_agent_to_session_agent(raw_agent)
            result = create_conversation(
                dev_repo_root,
                book_id=book_id,
                agent_key=agent_key,
                active_file=raw_active_file.strip(),
                title=raw_title if isinstance(raw_title, str) else None,
            )
        except ValueError as exc:
            return json_error("INVALID_PAYLOAD", str(exc), 400)
        return jsonify({"status": "success", **result}), 200

    @bp.post("/api/conversations/activate")
    def activate_conversation_route():
        payload = request.get_json(silent=True) or {}
        book_id, book_err = require_book_id(payload)
        if book_err:
            return book_err
        raw_agent = payload.get("agent", "")
        raw_conversation_id = payload.get("conversation_id", "")
        if not isinstance(raw_conversation_id, str) or not raw_conversation_id.strip():
            return json_error("INVALID_PAYLOAD", "conversation_id is required", 400)
        try:
            agent_key = route_agent_to_session_agent(raw_agent)
            result = activate_conversation(
                dev_repo_root,
                book_id=book_id,
                agent_key=agent_key,
                conversation_id=raw_conversation_id.strip(),
            )
        except ValueError:
            return json_error("NOT_FOUND", "conversation_not_found", 404)
        return jsonify({"status": "success", **result}), 200

    @bp.post("/api/conversations/rename")
    def rename_conversation_route():
        payload = request.get_json(silent=True) or {}
        book_id, book_err = require_book_id(payload)
        if book_err:
            return book_err
        raw_agent = payload.get("agent", "")
        raw_conversation_id = payload.get("conversation_id", "")
        raw_title = payload.get("title", "")
        if not isinstance(raw_conversation_id, str) or not raw_conversation_id.strip():
            return json_error("INVALID_PAYLOAD", "conversation_id is required", 400)
        if not isinstance(raw_title, str) or not raw_title.strip():
            return json_error("INVALID_PAYLOAD", "title is required", 400)
        try:
            agent_key = route_agent_to_session_agent(raw_agent)
            result = rename_conversation(
                dev_repo_root,
                book_id=book_id,
                agent_key=agent_key,
                conversation_id=raw_conversation_id.strip(),
                title=raw_title.strip(),
            )
        except ValueError:
            return json_error("NOT_FOUND", "conversation_not_found", 404)
        return jsonify({"status": "success", **result}), 200

    @bp.post("/api/conversations/archive")
    def archive_conversation_route():
        payload = request.get_json(silent=True) or {}
        book_id, book_err = require_book_id(payload)
        if book_err:
            return book_err
        raw_agent = payload.get("agent", "")
        raw_conversation_id = payload.get("conversation_id", "")
        if not isinstance(raw_conversation_id, str) or not raw_conversation_id.strip():
            return json_error("INVALID_PAYLOAD", "conversation_id is required", 400)
        try:
            agent_key = route_agent_to_session_agent(raw_agent)
            result = archive_conversation(
                dev_repo_root,
                book_id=book_id,
                agent_key=agent_key,
                conversation_id=raw_conversation_id.strip(),
            )
        except ValueError:
            return json_error("NOT_FOUND", "conversation_not_found", 404)
        return jsonify({"status": "success", **result}), 200

    @bp.post("/api/conversations/delete")
    def delete_conversation_route():
        payload = request.get_json(silent=True) or {}
        book_id, book_err = require_book_id(payload)
        if book_err:
            return book_err
        raw_agent = payload.get("agent", "")
        raw_conversation_id = payload.get("conversation_id", "")
        if not isinstance(raw_conversation_id, str) or not raw_conversation_id.strip():
            return json_error("INVALID_PAYLOAD", "conversation_id is required", 400)
        try:
            agent_key = route_agent_to_session_agent(raw_agent)
            result = delete_conversation(
                dev_repo_root,
                book_id=book_id,
                agent_key=agent_key,
                conversation_id=raw_conversation_id.strip(),
            )
        except ValueError:
            return json_error("NOT_FOUND", "conversation_not_found", 404)
        return jsonify({"status": "success", **result}), 200

    return bp
